add mypy to pre-commit

The code base currently has some `mypy` errors. Therefore, we can't simply add the `mypy` to the pre-commit hooks, because we wouldn't be able to commit due to the errors. 

I'll open some PRs to:

1. Fix `mypy` errors
2. Add `mypy` to pre-commit

`mypy` errors:

```
#mypy lightly tests
lightly/models/modules/masked_causal_vision_transformer.py: note: In member "forward" of class "MaskedCausalAttention":
lightly/models/modules/masked_causal_vision_transformer.py:21: error: Signature of "forward" incompatible with supertype "Attention"  [override]
lightly/models/modules/masked_causal_vision_transformer.py:21: note:      Superclass:
lightly/models/modules/masked_causal_vision_transformer.py:21: note:          def forward(self, x: Tensor, attn_mask: Tensor | None = ..., is_causal: bool = ...) -> Tensor
lightly/models/modules/masked_causal_vision_transformer.py:21: note:      Subclass:
lightly/models/modules/masked_causal_vision_transformer.py:21: note:          def forward(self, x: Tensor, mask: Tensor | None = ...) -> Tensor
lightly/models/modules/masked_causal_vision_transformer.py: note: In member "forward" of class "MaskedCausalBlock":
lightly/models/modules/masked_causal_vision_transformer.py:126: error: Signature of "forward" incompatible with supertype "Block"  [override]
lightly/models/modules/masked_causal_vision_transformer.py:126: note:      Superclass:
lightly/models/modules/masked_causal_vision_transformer.py:126: note:          def forward(self, x: Tensor, attn_mask: Tensor | None = ..., is_causal: bool = ...) -> Tensor
lightly/models/modules/masked_causal_vision_transformer.py:126: note:      Subclass:
lightly/models/modules/masked_causal_vision_transformer.py:126: note:          def forward(self, x: Tensor, mask: Tensor | None = ...) -> Tensor
lightly/models/modules/masked_causal_vision_transformer.py: note: In member "forward_features" of class "MaskedCausalVisionTransformer":
lightly/models/modules/masked_causal_vision_transformer.py:170: error: Signature of "forward_features" incompatible with supertype "VisionTransformer"  [override]
lightly/models/modules/masked_causal_vision_transformer.py:170: note:      Superclass:
lightly/models/modules/masked_causal_vision_transformer.py:170: note:          def forward_features(self, x: Tensor, attn_mask: Tensor | None = ..., is_causal: bool = ...) -> Tensor
lightly/models/modules/masked_causal_vision_transformer.py:170: note:      Subclass:
lightly/models/modules/masked_causal_vision_transformer.py:170: note:          def forward_features(self, x: Tensor, mask: Tensor | None = ...) -> Tensor
tests/loss/test_detcon_loss.py: note: In function "byol_nce_detcon":
tests/loss/test_detcon_loss.py:223: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[float64]]", variable has type "ndarray[tuple[int, int], dtype[float64]]")  [assignment]
Found 4 errors in 2 files (checked 530 source files)

```

Working on this. Implemented as a `repo: local` hook with `language: system` so it runs in the project's venv (`mirrors-mypy` won't work here — `torch` can't be listed as an `additional_dependency`).

The 4 errors listed in the issue are no longer present on master — looks like they've been fixed since this was filed. However, `mypy lightly tests` now surfaces 8 different errors across 4 files:

```
lightly/models/modules/ijepa_timm.py:94: error: Argument "norm_layer" to "Block" has incompatible type "Callable[..., Module]"; expected "type[Module]"  [arg-type]
lightly/models/modules/masked_autoencoder_timm.py:92: error: Argument "norm_layer" to "Block" has incompatible type "Callable[..., Module]"; expected "type[Module]"  [arg-type]
lightly/models/modules/masked_vision_transformer_timm.py:82: error: Statement is unreachable  [unreachable]
lightly/models/modules/masked_vision_transformer_timm.py:199: error: Argument 1 to "normal_" has incompatible type "Optional[Parameter]"; expected "Tensor"  [arg-type]
tests/loss/test_msn_loss.py:116: error: "Tensor" not callable  [operator]
tests/loss/test_msn_loss.py:120: error: Argument 1 to "any" has incompatible type "Union[bool, Any]"; expected "Tensor"  [arg-type]
tests/loss/test_msn_loss.py:138: error: "Tensor" not callable  [operator]
tests/loss/test_msn_loss.py:142: error: Argument 1 to "any" has incompatible type "Union[bool, Any]"; expected "Tensor"  [arg-type]
```

Would you like me to fix these in a separate PR first (matching your original plan), include them in the hook PR, or suppress them with targeted `# type: ignore` comments so the hook PR can land independently?

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -11,3 +11,12 @@ repos:
     - id: ruff-check
       args: [--fix]
     - id: ruff-format
+- repo: local
+  hooks:
+  - id: mypy
+    name: mypy
+    entry: mypy
+    language: system
+    types: [python]
+    pass_filenames: false
+    args: [lightly, tests]

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,3 +1,4 @@
+default_install_hook_types: [pre-commit, pre-push] # install both git hooks (mypy runs on pre-push)
 repos:
 - repo: https://github.com/pre-commit/pre-commit-hooks
   rev: v4.6.0
@@ -19,4 +20,5 @@ repos:
     language: system
     types: [python]
     pass_filenames: false
+    stages: [pre-push] # a full mypy run is too slow for every commit
     args: [lightly, tests]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -84,6 +84,7 @@ dev = [
   "types-python-dateutil",
   "types-toml",
   "types-requests",
+  "types-PyYAML",
   "nbformat",
   "jupytext"
 ]

--- a/tests/test_requirements.py
+++ b/tests/test_requirements.py
@@ -25,8 +25,9 @@ def test_requirements_base__openapi() -> None:
 
 
 def test_pre_commit_config__mypy_hook_config() -> None:
-    """Check mypy hook uses the project's installed environment and runs on
-    the full project rather than individual staged files.
+    """Check mypy hook uses the project's installed environment, runs on the
+    full project rather than individual staged files, and is scoped to the
+    pre-push stage (a full mypy run is too slow for every commit).
     """
     hook = _pre_commit_mypy_hook()
     assert hook is not None, "mypy hook not found in .pre-commit-config.yaml"
@@ -42,6 +43,10 @@ def test_pre_commit_config__mypy_hook_config() -> None:
     assert "lightly" in args and "tests" in args, (
         "mypy hook must pass 'lightly' and 'tests' as args — without them mypy "
         "runs with no targets and checks nothing"
+    )
+    assert "pre-push" in hook.get("stages", []), (
+        "mypy hook must be scoped to the pre-push stage so a full mypy run "
+        "does not slow down every commit"
     )
 
 

--- a/tests/test_requirements.py
+++ b/tests/test_requirements.py
@@ -1,7 +1,8 @@
 from pathlib import Path
-from typing import List, Set
+from typing import Any, Dict, List, Optional, Set, cast
 
 import toml
+import yaml
 
 
 def test_requirements_base__pyproject() -> None:
@@ -21,6 +22,38 @@ def test_requirements_base__openapi() -> None:
     }
     missing_deps = openapi - _requirements("base")
     assert not missing_deps
+
+
+def test_pre_commit_config__mypy_hook_config() -> None:
+    """Check mypy hook uses the project's installed environment and runs on
+    the full project rather than individual staged files.
+    """
+    hook = _pre_commit_mypy_hook()
+    assert hook is not None, "mypy hook not found in .pre-commit-config.yaml"
+    assert hook.get("language") == "system", (
+        "mypy hook must use language: system so it runs with the project's "
+        "installed torch and type stubs, not an isolated environment"
+    )
+    assert hook.get("pass_filenames") is False, (
+        "mypy hook must set pass_filenames: false — mypy needs to analyse the "
+        "whole project at once, not individual staged files"
+    )
+    args = hook.get("args", [])
+    assert "lightly" in args and "tests" in args, (
+        "mypy hook must pass 'lightly' and 'tests' as args — without them mypy "
+        "runs with no targets and checks nothing"
+    )
+
+
+def _pre_commit_mypy_hook() -> Optional[Dict[str, Any]]:
+    """Returns the mypy hook dict from .pre-commit-config.yaml."""
+    with open(".pre-commit-config.yaml") as f:
+        config = yaml.safe_load(f)
+    for repo in config["repos"]:
+        for hook in repo.get("hooks", []):
+            if hook["id"] == "mypy":
+                return cast(Dict[str, Any], hook)
+    return None
 
 
 def _pyproject(name: str) -> Set[str]:

